tests/test_grid_container.py

```diff
@@ -89,92 +86,103 @@ def _check(self, container: GridContainer[Dict[str, int]]) -> None:
         full_bbox = Box2I(Point2I(x=1, y=2), Extent2I(x=15, y=12))
         full_shape = Index2D(x=5, y=6)
-        full_builder: GridContainerBuilder[Dict[str, int]] = GridContainerBuilder(full_shape)
-        self._fill(full_builder)
-        full_container = full_builder.finish()
-        grid = UniformGrid(full_bbox, cell_size)
+        full_container = GridContainer(shape=full_shape)
+        self._fill(full_container)
+        grid = UniformGrid.from_bbox_cell_size(full_bbox, cell_size)
         self.assertEqual(grid.shape, full_shape)
-        # Subset with the orignal bounding box; should behave like a copy.
+
+        # Subset with the orignal bounding box; should behave like a deepcopy.
         subset_container_full = full_container.subset_overlapping(grid, full_bbox)
         self.assertEqual(subset_container_full.shape, full_container.shape)
         self.assertEqual(subset_container_full.offset, full_container.offset)
         self.assertEqual(list(subset_container_full), list(full_container))
+        subset_container_full[Index2D(x=2, y=2)] = {"x": -1, "y": -1}
+        self.assertNotEqual(list(subset_container_full), list(full_container))
+
         # Subset the full container with a nontrivial bbox.
         bbox_1 = Box2I(Point2I(x=6, y=4), Point2I(x=10, y=7))
         subset_container_1 = full_container.subset_overlapping(grid, bbox_1)
         self.assertEqual(subset_container_1.offset, Index2D(x=1, y=1))
         self.assertEqual(subset_container_1.shape, Index2D(x=3, y=2))
         union_bbox_1 = Box2I()
-        for v in subset_container_1:
-            cell_bbox = grid.bbox_of(Index2D(**v))
+        for v in subset_container_1.keys():
+            cell_bbox = grid.bbox_of(v)
             self.assertTrue(cell_bbox.overlaps(bbox_1))
             union_bbox_1.include(cell_bbox)
         self.assertTrue(union_bbox_1.contains(bbox_1))
         self._check(subset_container_1)
+
         # Subset the subset container with an even smaller bbox, to check the
         # case where the original offset is nonzero.
         bbox_2 = Box2I(Point2I(x=6, y=5), Point2I(x=7, y=5))
         subset_container_2 = subset_container_1.subset_overlapping(grid, bbox_2)
         self.assertEqual(subset_container_2.offset, Index2D(x=1, y=1))
         self.assertEqual(subset_container_2.shape, Index2D(x=2, y=1))
         union_bbox_2 = Box2I()
-        for v in subset_container_2:
-            cell_bbox = grid.bbox_of(Index2D(**v))
+        for v in subset_container_2.keys():
+            cell_bbox = grid.bbox_of(v)
             self.assertTrue(cell_bbox.overlaps(bbox_1))
             union_bbox_2.include(cell_bbox)
         self.assertTrue(union_bbox_2.contains(bbox_2))
         self._check(subset_container_2)
+
         # Subsetting the container by a bbox that isn't contained by the cells
-        # is an error.
-        with self.assertRaises(LengthError):
+        # raise a KeyError.
+        with self.assertRaises(KeyError):
```

Comment:
I think it might be useful to have bad bbox slices raise a slightly different exception; maybe `IndexError` instead?

Reply:
`KeyError` is raised naturally from these lines:
```python
for index in gc.indices():
    gc[index] = self[index]
```
Raising an `IndexError` would require to catch one expection and re-raise it as another. Should I do it that way instead?